### ISSUE TITLE
How to enable BTF support for ubuntu 20.04?

### ISSUE THREAD

**Issue by wwwzrb** [iovisor/bcc#2948]
I tried to enable BTF support by rebuilding kernel with CONFIG_DEBUG_INFO_BTF=y.

The specific configuration is shown below:

Kernel version:
![image](https://user-images.githubusercontent.com/16064432/83645652-4c096780-a5e5-11ea-8c87-0aa098da1760.png)

Clang and LLVM version:
![image](https://user-images.githubusercontent.com/16064432/83645813-7e1ac980-a5e5-11ea-916f-3c17b725b485.png)
![image](https://user-images.githubusercontent.com/16064432/83645862-8a9f2200-a5e5-11ea-8f40-9ebba166aab2.png)

And /sys/kernel/btf/vmlinux exist.
![image](https://user-images.githubusercontent.com/16064432/83646120-d651cb80-a5e5-11ea-9ed9-b260e5bab8e5.png)
bpftool btf dump file /sys/kernel/btf/vmlinux format c > vmlinux.h is OK.

**However, I encouter this problem:**
![image](https://user-images.githubusercontent.com/16064432/83646391-2761bf80-a5e6-11ea-8935-7b479b9e5c8c.png)

So actually opensnoop use kprobe and kretprobe to get the filled funcation arguments. KFUNC and KRETFUNC are not load successfully.

I don't what's going wrong. Is there anybody meet similar problems? Very gratefull if you have any suggestions!

**Comment by yonghong-song** [comment]
Could you make sure you have pahole >= 1.16 on your system?

**Comment by wwwzrb** [comment]
> Could you make sure you have pahole >= 1.16 on your system?

No, the pahole version is 1.15 in my system. I install pahole by warves package.

![image](https://user-images.githubusercontent.com/16064432/83657807-ee304c00-a5f3-11ea-803b-fe70bf8d6b19.png)

Is pahole the key problem?

**Comment by yonghong-song** [comment]
Yes, you need >= 1.16 pahole.

**Comment by wwwzrb** [comment]
> Yes, you need >= 1.16 pahole.

I rebuild my kernel with pahole 1.16 and also rebuild bcc. Now it works.

Pahole 1.16 solve the problem of "libbpf: bpf_prog_put is not found in vmulinux BTF"!
![image](https://user-images.githubusercontent.com/16064432/83719060-dfd04780-a668-11ea-9c3d-240a7072ce0d.png)

However, I encounter another problem. Do KFUNC and KRETFUNC support all kernel functions? I tried the opensnoop example and check whether KREFUNC for do_sys_open is loaded successfully. 

`is_support_kfunc = BPF.support_kfunc()
if is_support_kfunc:
    print("w/ KFUNC/KRETFUNC_PROBE SUPPORT")
    bpf_text += bpf_text_kfunc
else:
    print("w/o KFUNC/KRETFUNC_PROBE SUPPORT")
    bpf_text += bpf_text_kprobe`

Result shows that is_support_kfunc return a false value.
![image](https://user-images.githubusercontent.com/16064432/83719379-8e748800-a669-11ea-8514-0531d4aa96b3.png)

Or is this problem blame to not-upate-to-date kernel version?

I'm using kernel 5.4.34.

Thanks for your help!

**Comment by yonghong-song** [comment]
I think 5.4 does not support kfunc/kretfunc. You probably need 5.6. cc @olsajiri

**Comment by wwwzrb** [comment]
> I think 5.4 does not support kfunc/kretfunc. You probably need 5.6. cc @olsajiri

Got it. I'll have a try for newer version!

**Comment by wwwzrb** [comment]
> I think 5.4 does not support kfunc/kretfunc. You probably need 5.6. cc @olsajiri

Hey, I using kernel 5.7 now. It works but another new problem comes.

I'm tring to get the filled arguments after function returns. 

A simple example will be like this:

       KRETFUNC_PROBE(do_sys_open, int dfd, const char __user *filename, int flags, int mode, int ret)

Then I can get filename which may be filled after function returns.

**The problem is that KREFUNC_PROBE cannot deal type of pointer to struct. As attached below, struct sock * is unrecoginzed!**

![image](https://user-images.githubusercontent.com/16064432/83740414-61d06880-a689-11ea-8fee-261ebcf3b0b7.png)

**By the way, when I tried to include "vmlinux.h" instead of using kernel header, it causes errors of duplicate defination. Part reulsts are attached below.**

![image](https://user-images.githubusercontent.com/16064432/83740658-bbd12e00-a689-11ea-8246-84b8ac40f9ae.png)

**Comment by iammattcoleman** [comment]
Canonical has updated Ubuntu 20.04's 5.4 kernel to include BTF support. Installing the kernel 5.4 package from the focal-updates repository will fix that issue. Additionally, the `dwarves` package currently provides `pahole` 1.21.

Here is the last 20.04 kernel config that did not have BTF support:
* https://launchpad.net/ubuntu/+source/linux/5.4.0-91.102
* https://git.launchpad.net/~ubuntu-kernel/ubuntu/+source/linux/+git/focal/tree/debian.master/config/config.common.ubuntu?h=Ubuntu-5.4.0-91.102#n2322

The next version has it enabled:
* https://launchpad.net/ubuntu/+source/linux/5.4.0-92.103
* https://git.launchpad.net/~ubuntu-kernel/ubuntu/+source/linux/+git/focal/tree/debian.master/config/config.common.ubuntu?h=Ubuntu-5.4.0-92.103#n2322
